espnet/nets/pytorch_backend/transformer/decoder.py

```diff
@@ -150,12 +150,25 @@ def forward_one_step(self, tgt, tgt_mask, memory, cache=None):
         return y, new_cache
 
     # beam search API (see ScorerInterface)
-    def init_state(self, x=None):
-        """Get an initial state for decoding."""
-        return [None for i in range(len(self.decoders))]
-
     def score(self, ys, state, x):
         """Score."""
-        ys_mask = subsequent_mask(len(ys), device=x.device).unsqueeze(0)
-        logp, state = self.forward_one_step(ys.unsqueeze(0), ys_mask, x.unsqueeze(0), cache=state)
-        return logp.squeeze(0), state
+        if ys.dim() == 1:
+            ys_mask = subsequent_mask(len(ys), device=x.device).unsqueeze(0)
+            logp, state = self.forward_one_step(ys.unsqueeze(0), ys_mask, x.unsqueeze(0), cache=state)
+            return logp.squeeze(0), state
+
+        # merge states
+        n_batch = len(ys)
+        n_layers = len(self.decoders)
+        if state[0] is None:
+            batch_state = None
+        else:
+            batch_state = [torch.stack([state[b][l] for b in range(n_batch)]) for l in range(n_layers)]
+
+        # batch decoding
+        ys_mask = subsequent_mask(ys.size(-1), device=x.device).unsqueeze(0)
+        logp, state = self.forward_one_step(ys, ys_mask, x, cache=batch_state)
+
+        # split states
+        state_list = [[state[l][b] for l in range(n_layers)] for b in range(n_batch)]
```

Comment:
Here the shape of `state` may be confused.
In the first line, the index is [batch, layer] but in the last line [layer, batch].
It is better to add shape info as a comment or add a docstring.

Reply:
OK